File: DataBaseRoomApp/jroomapp/src/main/java/com/woowrale/jroomapp/di/modules/MainModule.java
package com.woowrale.jroomapp.di.modules;

import android.content.Context;

import com.woowrale.jroomapp.MainApplication;
import com.woowrale.jroomapp.data.database.ContactDatabase;
import com.woowrale.jroomapp.data.datasource.ContactDataSource;
import com.woowrale.jroomapp.data.datasource.LocalDataSource;
import com.woowrale.jroomapp.data.repository.LocalRepository;
import com.woowrale.jroomapp.di.threads.JobThread;
import com.woowrale.jroomapp.di.threads.UIThread;
import com.woowrale.jroomapp.usecases.threads.JobScheduler;
import com.woowrale.jroomapp.usecases.threads.UIScheduler;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class MainModule {

    private final MainApplication mainApplication;

    public MainModule(MainApplication mainApplication) {
        this.mainApplication = mainApplication;
    }

    @Provides
    @Singleton
    public Context provideContext(){
        return mainApplication;
    }

    @Provides
    @Singleton
    public ContactDatabase provideDataBase(Context context){
        return ContactDatabase.build(context);
    }

    @Provides
    @Singleton
    public LocalDataSource provideContactDataSource(ContactDatabase contactDatabase) {
        return new ContactDataSource(contactDatabase);
    }

    @Provides
    @Singleton
    public LocalRepository provideRepository(LocalDataSource localDataSource) {
        return new LocalRepository(localDataSource);
    }

    @Provides
    @Singleton
    public JobScheduler provideJobScheduler(JobThread jobThread){
        return jobThread;
    }

    @Provides
    @Singleton
    public UIScheduler provideUIScheduler(UIThread uiThread){
        return uiThread;
    }

}

File: DataBaseRoomApp/kroomapp/src/main/java/com/woowrale/kroomapp/di/modules/MainModule.kt
package com.woowrale.kroomapp.di.modules

import android.content.Context
import com.woowrale.kroomapp.MainApplication
import com.woowrale.kroomapp.data.database.ContactDatabase
import com.woowrale.kroomapp.data.datasource.ContactDataSource
import com.woowrale.kroomapp.data.datasource.LocalDataSource
import com.woowrale.kroomapp.data.repository.LocalRepository
import com.woowrale.kroomapp.di.threads.JobThread
import com.woowrale.kroomapp.di.threads.UIThread
import com.woowrale.kroomapp.usecases.threads.JobScheduler
import com.woowrale.kroomapp.usecases.threads.UIScheduler
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class MainModule constructor(private val mApplication: MainApplication) {

    @Provides
    @Singleton
    fun provideContext(): Context {
        return mApplication
    }

    @Provides
    @Singleton
    fun provideDataBase(context: Context): ContactDatabase {
        return ContactDatabase.build(context)
    }

    @Provides
    fun provideContactDataSource(contactDatabase: ContactDatabase): LocalDataSource {
        return ContactDataSource(contactDatabase)
    }

    @Provides
    fun provideRepository(localDataSource: LocalDataSource): LocalRepository {
        return LocalRepository(localDataSource)
    }

    @Provides
    fun provideJobScheduler(jobThread: JobThread): JobScheduler {
        return jobThread
    }

    @Provides
    fun provideUIScheduler(uiThread: UIThread): UIScheduler {
        return uiThread
    }
}
